jax/lax/lax.py

```diff
@@ -573,6 +572,94 @@ def conv_general_dilated(
+  contains flattened image patches, so instead a single `"C"` dimension
+  represents, for example, three dimensions `"chw"` collapsed. The order of
+  these dimensions is `"c" + ''.join(c for c in rhs_spec if c not in 'OI')`,
+  where `rhs_spec == dimension_numbers[1]`, and the size of this `"C"`
+  dimension is therefore the size of each patch, i.e.
+  `np.prod(filter_shape) * lhs.shape[lhs_spec.index('C')]`, where
+  `lhs_spec == dimension_numbers[0]`.
+
+  Docstring below adapted from `jax.lax.conv_general_dilated`.
+
+  See Also:
+    https://www.tensorflow.org/xla/operation_semantics#conv_convolution
+
+  Args:
+    lhs: a rank `n+2` dimensional input array.
+    filter_shape: a sequence of `n` integers, representing the receptive window
+      spatial shape in the order as specified in
+      `rhs_spec = dimension_numbers[1]`.
+    window_strides: a sequence of `n` integers, representing the inter-window
+      strides.
+    padding: either the string `'SAME'`, the string `'VALID'`, or a sequence of
+      `n` `(low, high)` integer pairs that give the padding to apply before and
+      after each spatial dimension.
+    lhs_dilation: `None`, or a sequence of `n` integers, giving the
+      dilation factor to apply in each spatial dimension of `lhs`. LHS dilation
+      is also known as transposed convolution.
+    rhs_dilation: `None`, or a sequence of `n` integers, giving the
+      dilation factor to apply in each spatial dimension of `rhs`. RHS dilation
+      is also known as atrous convolution.
+    dimension_numbers: either `None`, or a 3-tuple
+      `(lhs_spec, rhs_spec, out_spec)`, where each element is a string
+      of length `n+2`. `None` defaults to `("NCHWD..., OIHWD..., NCHWD...")`.
+    precision: Optional. Either ``None``, which means the default precision for
+      the backend, or a ``lax.Precision`` enum value (``Precision.DEFAULT``,
+      ``Precision.HIGH`` or ``Precision.HIGHEST``).
+
+  Returns:
+    A rank `n+2` array containing the flattened image patches in the output
+    channel (`"C"`) dimension. For example if
+    `dimension_numbers = ("NcHW", "OIwh", "CNHW")`, the output has dimension
+    numbers `"CNHW" = "{cwh}NHW"`, with the size of dimension `"C"` equal to
+    the size of each patch
+    (`np.prod(filter_shape) * lhs.shape[lhs_spec.index('C')]`).
+
+  """
+  filter_shape = tuple(filter_shape)
+  dimension_numbers = conv_dimension_numbers(lhs.shape, (1, 1) + filter_shape,
+                                             dimension_numbers)
+
+  lhs_spec, rhs_spec, out_spec = dimension_numbers
+
+  spatial_size = np.prod(filter_shape)
+  n_channels = lhs.shape[lhs_spec[1]]
+
+  # Move separate `lhs` spatial locations into separate `rhs` channels.
+  rhs = _eye(lhs.dtype, (spatial_size,) * 2, 0).reshape(filter_shape * 2)
+
+  rhs = rhs.reshape((spatial_size, 1) + filter_shape)
+  rhs = np.tile(rhs, (n_channels,) + (1,) * (rhs.ndim - 1))
+  rhs = np.moveaxis(rhs, (0, 1), (rhs_spec[0], rhs_spec[1]))
```

Comment:
In the post-omnistaging future, it seems like it would probably be wiser to use `jnp` functions for generating these arrays rather than using `np`?

Reply:
Good point, I didn't notice I was using `onp`! Given that I can't import `jnp` here since it itself depends on `lax` and this would lead to a circular dependency, where would you advice me to put this function? I could put it in a separate file `lax/lax_other.py` and import in `lax/__init__.py` from there. Lmk what you think, thanks!

Reply:
I would guess we could also get away with a local import inside the function, but `lax/lax_other.py` is probably a better idea!

Reply:
Moved to lax_other!